keyvi/include/keyvi/dictionary/dictionary_merger.h

```diff
@@ -197,20 +243,69 @@ class DictionaryMerger final {
 
     TRACE("finished iterating, do final compile.");
 
-    generator->CloseFeeding();
+    generator_->CloseFeeding();
+  }
+
+  void Write(std::ostream& stream) {
+    if (!generator_) {
+      throw merger_exception("not merged yet");
+    }
 
-    generator->SetManifestFromString(manifest_);
-    generator->WriteToFile(filename);
+    generator_->Write(stream);
   }
 
+  void WriteToFile(const std::string& filename) {
+    if (!generator_) {
+      throw merger_exception("not merged yet");
+    }
+    generator_->WriteToFile(filename);
+  }
+
+  const MergeStats& GetStats() const { return stats_; }
+
  private:
+  typename GeneratorAdapter::AdapterPtr generator_;
   bool append_merge_ = false;
   std::vector<fsa::automata_t> dicts_to_merge_;
+  std::vector<std::vector<std::string>> deleted_keys_;
   std::vector<std::string> inputFiles_;
   std::priority_queue<SegmentIterator> segments_pqueue_;
-
-  fsa::internal::IValueStoreWriter::vs_param_t params_;
+  keyvi::util::parameters_t params_;
   std::string manifest_ = std::string();
+  MergeStats stats_;
+
+  /**
+   * Load a file with deleted keys if it exists
+   */
+  std::vector<std::string> TryLoadDeletedKeys(const std::string& filename) {
+    std::vector<std::string> deleted_keys;
+    boost::filesystem::path deleted_keys_file{filename};
+    deleted_keys_file += ".dk";
+
+    TRACE("check for deleted keys file: %s", deleted_keys_file.string().c_str());
+    std::ifstream deleted_keys_stream(deleted_keys_file.string(), std::ios::binary);
+
+    if (deleted_keys_stream.good()) {
+      TRACE("found deleted keys file");
+
+      {
+        // reads the buffer as 1 big chunk, could be improved
+        // msgpack v2.x provides a better interface (visitor)
+        std::stringstream buffer;
+        buffer << deleted_keys_stream.rdbuf();
+
+        msgpack::unpacked unpacked_object;
```

Comment:
Also the format of file is defined somewhere else, will be great that writer and reader parts would live together in centralized place.